core/Network/TLS/Handshake/Server.hs

```diff
@@ -152,16 +152,17 @@ handshakeServerWith sparams ctx clientHello@(ClientHello clientVersion _ clientS
            TLS12 -> let -- Build a list of all signature algorithms with at least
                         -- one hash algorithm in common between client and server.
                         -- May contain duplicates, as it is only used for `elem`.
-                        possibleSigAlgs = map snd (hashAndSignaturesInCommon ctx exts)
+                        possibleHashSigAlgs = hashAndSignaturesInCommon ctx exts
+                        possibleSigAlgs = map snd possibleHashSigAlgs
 
                         -- Check that a candidate cipher with a signature requiring
                         -- a hash will have at least one hash available.  This avoids
                         -- a failure later in 'decideHash'.
                         hasSigningRequirements =
                             case cipherKeyExchange cipher of
-                                CipherKeyExchange_DHE_RSA      -> SignatureRSA   `elem` possibleSigAlgs
+                                CipherKeyExchange_DHE_RSA      -> isJust $ find (RSA `signatureCompatible`) possibleHashSigAlgs
```

Comment:
- signatureCompatible can be used for DSS and ECDSA too.
- elemBy is `Data.List.any`